focus-android/app/src/main/java/org/mozilla/focus/cookiebannerreducer/CookieBannerReducerMiddleware.kt

```diff
@@ -0,0 +1,203 @@
+
+    private fun handleCookieBannerToggle(
+        action: CookieBannerReducerAction.ToggleCookieBannerException,
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        ioScope.launch {
+            if (action.isCookieBannerHandlingExceptionEnabled) {
+                cookieBannersStorage.removeException(currentTab.content.url, true)
+                CookieBanner.exceptionRemoved.record(NoExtras())
+                context.store.dispatch(
+                    CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                        CookieBannerReducerStatus.NoException,
+                    ),
+                )
+            } else {
+                clearSiteData()
+                cookieBannersStorage.addPersistentExceptionInPrivateMode(currentTab.content.url)
+                CookieBanner.exceptionAdded.record(NoExtras())
+                context.store.dispatch(
+                    CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                        CookieBannerReducerStatus.HasException,
+                    ),
+                )
+            }
+            appContext.components.sessionUseCases.reload()
+        }
+    }
+
+    private fun reportSite(
+        action: CookieBannerReducerAction.RequestReportSite,
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        CookieBanner.reportSiteDomain.set(action.siteToReport)
+        CoroutineScope(Dispatchers.IO).launch {
+            Pings.cookieBannerReportSite.submit()
+            context.store.dispatch(
+                CookieBannerReducerAction.ShowSnackBarForSiteToReport(
+                    true,
+                ),
+            )
+            context.store.dispatch(
+                CookieBannerReducerAction.UpdateCookieBannerReducerStatus(
+                    CookieBannerReducerStatus.CookieBannerUnsupportedSiteRequestWasSubmitted,
+                ),
+            )
+        }
+    }
+
+    private fun initCookieBannerReducer(
+        context: MiddlewareContext<CookieBannerReducerState, CookieBannerReducerAction>,
+    ) {
+        val shouldShowCookieBannerItem = shouldShowCookieBannerReducerItem()
+        context.store.dispatch(
+            CookieBannerReducerAction.UpdateCookieBannerReducerVisibility(
+                shouldShowCookieBannerItem = shouldShowCookieBannerItem,
+            ),
+        )
+
+        if (!shouldShowCookieBannerItem) {
+            return
+        }
+        ioScope.launch {
+            val hasException =
+                cookieBannersStorage.hasException(currentTab.content.url, true)
+            withContext(Dispatchers.Main) {
+                if (!hasException) {
+                    callHasCookieBannerRuleForSession(context)
+                } else {
+                    showExceptionStatus(context, true)
+                }
+            }
+        }
+    }
+
+    private fun callHasCookieBannerRuleForSession(
```

Comment:
nit: I think we need to improve the function to be more descriptive, to indicate what is doing without telling the name of the function that it's calling internally. Maybe `showUnsupportedSiteIfNeeded`.

Reply:
Done.